marlowe-actus/src/Language/Marlowe/ACTUS/Generator.hs

```diff
@@ -57,101 +55,90 @@ maxPseudoDecimalValue = 100000000000000
-            , cashPaymentDay     = paymentDay d
-            , cashCalculationDay = calculationDay d
-            , cashEvent          = ev
-            , amount             = pff
-            , currency           = "ada"
-            }
-    in
-        genCashflow <$> L.zip states payoffs
+    
 
 genStaticContract :: ContractTerms -> Contract
 genStaticContract terms =
     let
         cfs = genProjectedCashflows terms
-        gen CashFlow{..} = invoice "party" "counterparty" (Constant $ round amount) (Slot $ dayToSlotNumber cashPaymentDay)
+        gen CashFlow{..} = 
+            if amount > 0.0 
+                then invoice "party" "counterparty" (Constant $ round amount) (Slot $ dayToSlotNumber cashPaymentDay)
+                else invoice "counterparty" "party" (Constant $ round $ -amount) (Slot $ dayToSlotNumber cashPaymentDay)
     in foldl (flip gen) Close cfs
 
 
 genFsContract :: ContractTerms -> Contract
 genFsContract terms =
     let
+        postProcess cont = 
+            let ctr = constraints terms
+                toAssert = genZeroRiskAssertions terms <$> (assertions =<< maybeToList ctr)
+                compose = appEndo . mconcat . map Endo
+            in compose toAssert cont
+
         payoffAt t = ValueId $ fromString $ "payoff_" ++ show t
         schedCfs = genProjectedCashflows terms
         schedEvents = cashEvent <$> schedCfs
         schedDates = Slot . dayToSlotNumber . cashPaymentDay <$> schedCfs
+        previousDates = ([ct_SD terms] ++ (cashCalculationDay <$> schedCfs))
         cfsDirections = amount <$> schedCfs
-        gen :: (CashFlow, EventType, Slot, Double, Integer) -> Contract -> Contract
-        gen (cf, ev, date, r, t) cont = inquiryFs ev ("_" ++ show t) date "oracle"
-            $ stateTransitionFs ev terms t (cashCalculationDay cf)
-            $ Let (payoffAt t) (payoffFs ev terms t (cashCalculationDay cf))
-            $ if r > 0.0    then invoice "party" "counterparty" (UseValue $ payoffAt t) date cont
-                            else invoice "counterparty" "party" (NegValue $ UseValue $ payoffAt t) date cont
-        scheduleAcc = foldr gen Close $ L.zip5 schedCfs schedEvents schedDates cfsDirections [1..]
+        ctx = context <$> constraints terms
+        
+        gen :: (CashFlow, Day, EventType, Slot, Double, Integer) -> Contract -> Contract
+        gen (cf, prevDate, ev, date, r, t) cont = 
+            inquiryFs ev terms ("_" ++ show t) date "oracle" ctx
+            $ stateTransitionFs ev terms t prevDate (cashCalculationDay cf)
+            $ Let (payoffAt t) (fromMaybe (constnt 0.0) pof)
+            $ if (isNothing pof) then cont
+              else if  r > 0.0   then invoice "party" "counterparty" (UseValue $ payoffAt t) date cont
+              else                    invoice "counterparty" "party" (NegValue $ UseValue $ payoffAt t) date cont
+            where pof = (payoffFs ev terms t (t - 1) prevDate (cashCalculationDay cf))
```

Comment:
POF has to be nullable. This code us not a very good quality - I agree.

All it really does is fold of all state transition. And applying POF to the result. If POF is nothing - I skip Pay -> Deposit in Marlowe. 

I can’t use 0.0 value because it could be non-zero for ACTUS-FS version.

My code here suffers from trying to be ACTUS-F and ACTUS-FS at the same time.

Reply:
but I mean as far as I can see you only use `pof` here as a boolean encoded into a `Maybe`

Reply:
@shmish111 I use it directly in genProjectedCashflows